drivers/serial/uart_nrfx_uarte.c

```diff
@@ -251,6 +307,453 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
 	return 0;
 }
 
+
+static int uarte_nrfx_err_check(struct device *dev)
+{
+	u32_t error = 0U;
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+
+	if (nrf_uarte_event_check(uarte, NRF_UARTE_EVENT_ERROR)) {
+		/* register bitfields maps to the defines in uart.h */
+		error = nrf_uarte_errorsrc_get_and_clear(uarte);
+	}
+
+	return error;
+}
+
+#ifdef CONFIG_UART_ASYNC_API
+
+static int uarte_nrfx_tx(struct device *dev, const u8_t *buf, size_t len,
+			 u32_t timeout)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	const struct uarte_nrfx_config *cfg = get_dev_config(dev);
+
+	if (data->async->tx) {
+		return -EBUSY;
+	}
+	data->async->tx = buf;
+	data->async->tx_len = len;
+	data->async->tx_cur_idx = 0;
+	nrf_uarte_tx_buffer_set(cfg->uarte_regs, buf, len);
+	nrf_uarte_event_clear(cfg->uarte_regs, NRF_UARTE_EVENT_TXSTARTED);
+	nrf_uarte_task_trigger(cfg->uarte_regs, NRF_UARTE_TASK_STARTTX);
+	if (cfg->uarte_regs->CONFIG & UARTE_CONFIG_HWFC_Msk) {
+		k_delayed_work_submit(&data->async->tx_timeout_work, timeout);
+	}
+	return 0;
+}
+
+#define INTERRUPTS_MASK							       \
+	(NRF_UARTE_INT_ENDRX_MASK | NRF_UARTE_INT_RXSTARTED_MASK |	       \
+	 NRF_UARTE_INT_ERROR_MASK | NRF_UARTE_INT_ENDTX_MASK |		       \
+	 NRF_UARTE_INT_TXSTOPPED_MASK | NRF_UARTE_INT_RXTO_MASK)
+
+static void user_callback(struct device *dev, struct uart_event *evt)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+
+	if (data->async->evt_handler) {
+		data->async->evt_handler(evt, data->async->user_data);
+	}
+}
+
+static void rx_timeout(struct device *dev)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	u32_t read;
+
+	if (IS_ENABLED(CONFIG_UARTE_NRF_HW_ASYNC) && data->async->hw_async) {
```

Comment:
This condition should be wrapped in an `inline` function (e.g. `hw_rx_counting_enabled`) and called wherever there is a need to check if the hardware counting is active. This way all the parts of code that appear to be dead have chance to be optimized out. Right now this won't happen, for example, to the `if` block related to the `RXDRDY` event in `uarte_nrfx_isr`.

BTW `hw_async` is very cryptic.